### ISSUE TITLE
My beeves with pixel_shuffle as named and implemented in torch, and some un–asked-for suggestions

### ISSUE THREAD

**Issue by celoyd** [celoyd/potato#69]
It’s Sunday night. Let’s resurrect something previously discarded from a code comment.

The pixel_shuffle operation was first published in [the ESPCN paper](https://arxiv.org/abs/1609.05158) (Shi et al. 2016), and I first saw it in [_Learning to See in the Dark_](https://arxiv.org/abs/1805.01934) (Chen et al. 2018), a paper I bring up whenever I can because it was the first CV ML paper I ever respected. It wouldn’t surprise me to learn that people working with Bayer arrays already had the concept in some other form – that might be an interesting literature search. In any case, I have it via ESPCN. There it’s called shuffling, and specifically _periodic shuffling_, which is at least one notch better than what it’s now called. That brings us to my first problem:

### That’s not what shuffling means

In card games, in general use, and specifically in computer science, to shuffle usually means to randomize an order. There are exceptions but they are exceptions. “Pixel shuffling” is not randomizing anything – it’s deterministic, necessarily and by design – so calling it a shuffle is confusing.

I’m writing a pansharpener called Potato. Names don’t have to be descriptions. But I think it’s a problem when a name sounds specifically, plausibly descriptive of something that it’s not.

### Which one is the inverse one?

Quick, does shuffling increase the HW size and decrease the C size, or is that unshuffling? Every time I don’t have to think about this for a week, I forget.

Shuffling cards is a deck→deck operation, not deal→deck or deck→deal. The only way to orient the arrow of time in the abstract idea of shuffling is that it’s the direction in which entropy increases, but this does not apply to “pixel shuffling”, because it’s not a shuffle, and so there is no natural forward/inverse polarity for the terms pixel_\[un\]shuffle. You just have to know the convention. It’s like if we called deflating a balloon air_scramble and inflating it air_unscramble. The terms should be more accurate but they should also be more evocative, given it’s a visualizable action.

Whenever the _shuffling is shallowing and widening; unshuffling is deepening and narrowing_ convention arose seems to be after the ESPCN paper, incidentally; it calls both directions shuffling. This suggests that its authors were using it as a sort of loose placeholder term for tricky re-ordering. (The only other non-random “shuffle” I can think of is what’s better called [the _bit interleaving_ in Z-order](https://en.wikipedia.org/wiki/Z-order_curve#Efficiently_building_quadtrees_and_octrees), which I assume inspired ESPCN’s trick and is the root of the name.) I choose to believe that the ESPCN authors are horrified by what the term became.

## It’s not, it doesn’t um, when the, if you, it can’t, compose the way, if you call it, if you make it recursive, the un-nesting is um, it isn’t, what do you call non-commutativity but if the, when you don’t fast-forward, it’s different, you know?

Okay, I don’t know group theory or whatever, so I don’t know the word for it. Here’s what it looks like:

```python
>>> import torch
>>> from torch.nn import functional as F
>>> img = torch.rand(1, 8, 64, 64)
>>> unshuffled4 = F.pixel_unshuffle(img, 4)
>>> unshuffled2n2 = F.pixel_unshuffle(F.pixel_unshuffle(img, 2), 2)
>>> (unshuffled4 - unshuffled2n2).abs().mean()
tensor(0.2917)
```

The {up,down}scale_factor argument to pixel_\[un\]shuffle looks like it might be (equivalent to) an iteration count of a base operation, and it could be, but it’s not. There are an infinite number of completely reasonable functions such that $`f(a, 2b) \neq f(f(a, b), b)`$ but it’s still annoying if your metal model was that _b_ is essentially a counter.

Here’s an example of where this could bite you. Imagine you have some image data that you want shuffled to two lower spatial resolutions. You might assume that you can bounce freely among levels of shuffling however seems most clear and convenient, perhaps like so:

```python
quarter = F.pixel_unshuffle(original, 4)
...
half = F.pixel_shuffle(quarter, 2)
```

And then you would have to spend an hour debugging your dang pansharpening model when for some reason it’s not converging properly, only to find, while checking what seemed like obvious assumptions, that pixel_unshuffle does not work like this, even though it could and seems like it ought to.

_Surprise!_ This is exactly what happened to a guy I know.

### Pile and tile

The names `pile` (for the deepening and narrowing operation) and `tile` (for its shallowing and widening opposite) don’t have strongly confusing associations and are clear about their directions. To pile things is to stack them and to tile things is to lay them out in an ordered way. They do not fully define, but at least they describe.

I think “space to depth” and “depth to space” are reasonable names, and I certainly encourage their use over you-know-what, but they’re a little long.

### Iterative version

These functions, implemented with `einops.rearrange`, satisfy the general definition of pixel_\[un\]shuffle ($`(\star, C, H \times r, W \times r) \leftrightarrow (\star, C \times r^2, H, W)`$) but do the ordering in such a way that `tile(pile(pile(x, 2), 2), 4) = x` and so on. They are unchecked for errors like odd dimensions.

```python
def tile(x, factor=2):
    times = int(torch.round(torch.log2(torch.tensor(factor))))
    for _ in range(times):
        x = rearrange(x, "... (c two dos) h w -> ... c (two h) (dos w)", two=2, dos=2)
    return x

def pile(x, factor=2):
    times = int(torch.round(torch.log2(torch.tensor(factor))))
    for _ in range(times):
        x = rearrange(x, "... c (two h) (dos w) -> ... (c two dos) h w", two=2, dos=2)
    return x
```

We keep the factor argument’s meaning. (There are at least two other reasonable things it could mean: number of iterations, i.e. the `times` in the above; or the _x_×_y_ area of pixels in the areas mapped to points, equivalent to the square of how it’s defined now, or to the factor of increase in the depth.)

I suspect, but have not proven, that einops cannot implement this in a single pass. I am a big fan of einops but it does seem stretched away from its core purpose here. A single-step version in einops (or perhaps in pure torch indexing?) would of course be preferable, all else being equal, but I am confident that an extra `rearrange` or two is unlikely to matter for practical purposes, and so I consider this a drop-in replacement.
